### PR TITLE
Add vocabulary object. Model loading infers RNN size.

### PR DESCRIPTION
**Pull request by annaproxy**
#26 Vocabulary object implemented. Vocabulary is assumed to be loaded from a text file. Multiple models can share the same vocabulary. It should therefore be sorted, so no weird mismatches happen. 
It is checked during loading of a model whether vocabulary is the same size.

#31 Model loading now inferns the RNN size from the current architecture. For future architectures, such as multiple layers, we should make sure this still happens.

### CODE REVIEW

File: app/ml_models/rnn/data_tools.py [hunk 1]
```diff
@@ -1,34 +1,46 @@
...
         self,
         prefix: str = "../../../data/",
-        filename_dataset: str = "straattaal.txt",
+        filename_datasets: List[str] = ["straattaal.txt"],
+        vocabulary: Vocabulary = None,
         filename_vocab: str = "vocabulary.txt",
     ):
-        """Initialize a WordLevelDataset object.
+        """
+        Initialize a WordLevelDataset object and accompanying Vocabulary object if necessary.
```
Comment (Sasafrass) [inline app/ml_models/rnn/data_tools.py:19]: Nit, officially google docstring is this if I'm not mistaken though I might as well be:
```suggestion
        """Initialize a WordLevelDataset object and accompanying Vocabulary object if necessary.
```

File: app/ml_models/rnn/data_tools.py [hunk 2]
```diff
@@ -1,34 +1,46 @@
...
 
         Args:
             prefix: The prefix to the folder containing the dataset.
-            filename_dataset: Full filename of dataset to be appended to the prefix.
-            filename_vocab: Full filename of vocabulary to be appended to the prefix.
+            filename_datasets: A list of full filenames of datasets to be appended to the prefix.
+            vocabulary: Optional Vocabulary object (see vocabulary.py). If set to None, vocabulary will be loaded from filename_vocab.
+            filename_vocab: Optional, only necessary if the vocabulary argument is None. Full filename of vocabulary to be appended to the prefix.
+
```
Comment (Sasafrass) [inline app/ml_models/rnn/data_tools.py:26]: nit
```suggestion
            vocabulary: Optional Vocabulary object (see vocabulary.py). If set to None, vocabulary will be loaded from filename_vocab.
            filename_vocab: Optional, only necessary if the vocabulary argument is None. Full filename of vocabulary to be appended to the prefix.
```

Comment (Sasafrass) [reply app/ml_models/rnn/data_tools.py:26]: Wait does it remove the empty line in my suggestion?

File: app/ml_models/rnn/loaded_rnn_model.py [hunk 3]
```diff
@@ -1,24 +1,46 @@
...
 
-def load_model(model_name: str = "2021_straattaal_epoch100.pt", device: str = "cpu"):
+def load_model(
+    filename_model: str = "2021_straattaal_epoch100.pt",
+    filename_vocab: str = "vocabulary.txt",
+    device: str = "cpu",
+):
     """
+    From a given path for filename of the model and the vocabulary, loads a model and vocabulary for inference.
+    The vocabulary size should be the same as the model vocabulary size, otherwise an error is thrown.
```
Comment (Sasafrass) [inline app/ml_models/rnn/loaded_rnn_model.py:16]: nice

File: app/ml_models/rnn/loaded_rnn_model.py [hunk 4]
```diff
@@ -1,24 +1,46 @@
...
+
+    v = Vocabulary()
+    v.load(prefix=".", filename_vocab=path_vocab)
+
+    checkpoint = torch.load(path_model, map_location=torch.device(device))
+    checkpoint_model = checkpoint["model_state_dict"]
+    model_vocab_size, embedding_size = checkpoint_model["_embedding.weight"].shape
+
+    # Will work for single-layer RNNs for now.
+    hidden_size = checkpoint_model["lstm.bias_hh_l0"].shape[0]
```
Comment (Sasafrass) [inline app/ml_models/rnn/loaded_rnn_model.py:35]: Let's open an issue for making this more generic in the future?

File: app/ml_models/rnn/vocabulary.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        filename_source: str = "dutch.txt", # TODO also fread from multiple files if necessary
+        filename_destination: str = "dutch_vocab.txt",
+        overwrite: bool = False,
+        lower: bool = True,
+    ):
+        """
+        Initializes a Vocabulary object by building it from a source (text) file containing all possible characters.
+
+        Args:
+            prefix: The prefix to the folder containing the source file
```
Comment (Sasafrass) [inline app/ml_models/rnn/vocabulary.py:17]:
```suggestion
            prefix: The prefix to the folder containing the source file.
```

File: app/ml_models/rnn/vocabulary.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                chars = set(f.read().lower())
+        chars.difference_update(set('\n\t"'))
+
+        # Sort the vocabulary
+        chars = "".join(sorted(chars))
+
+        with open(filename_destination, "w", encoding="utf-8") as f:
+            f.write(chars)
+
+        # TODO Paths, paths, I dont like them
```
Comment (Sasafrass) [inline app/ml_models/rnn/vocabulary.py:42]: RT

**Review by Sasafrass — APPROVED** [review]
LVGTM. Not a lot of comments - at least no blocking comments, so I'll approve so that you can continue with merging if you wish haha

File: app/ml_models/rnn/data_tools.py (hunk 2 above)
Comment (annaproxy) [reply app/ml_models/rnn/data_tools.py:25]: I will commit anyway, to give you a green block

File: app/ml_models/rnn/loaded_rnn_model.py (hunk 4 above)
Comment (annaproxy) [reply app/ml_models/rnn/loaded_rnn_model.py:35]: #44

Comment (annaproxy) [reply app/ml_models/rnn/loaded_rnn_model.py:35]: #44

**Pull request merged**